src/librustc/mir/mod.rs

```diff
@@ -1784,19 +1790,8 @@ pub enum StatementKind<'tcx> {
     /// (The starting point(s) arise implicitly from borrows.)
     EndRegion(region::Scope),
 
-    /// Encodes a user's type ascription. These need to be preserved
-    /// intact so that NLL can respect them. For example:
-    ///
-    ///     let a: T = y;
-    ///
-    /// The effect of this annotation is to relate the type `T_y` of the place `y`
-    /// to the user-given type `T`. The effect depends on the specified variance:
-    ///
-    /// - `Covariant` -- requires that `T_y <: T`
-    /// - `Contravariant` -- requires that `T_y :> T`
-    /// - `Invariant` -- requires that `T_y == T`
-    /// - `Bivariant` -- no effect
-    AscribeUserType(Place<'tcx>, ty::Variance, Box<UserTypeProjection<'tcx>>),
+    /// This represents the insertion of a user type ascription to be asserted.
+    AscribeUserType(UserTypeAscription),
```

Comment:
FWIW, I think I would make the `UserTypeAscription` only refer to the `UserTypeProjection<'tcx>` -- or, hmm, I guess that only makes sense relative to the place, huh? 

In a way, I think the thing we want in the array is probably the "kernel type", not the projection parts.

Reply:
Resolving this as all the code has changed.